From 10806c2f732e1ede40d9c938b4f4b4fabb256660 Mon Sep 17 00:00:00 2001
From: Peter Cai <peter@typeblog.net>
Date: Tue, 11 Aug 2026 08:49:13 +0000
Subject: [PATCH] Implement a persistent property to override the default
 primary camera (0)

Change-Id: I49b45d00bf71d7932591b3516d49a680e1b6568b

[Port to A16] cameraIds swap moved into sortCameraIds() tail so it covers both getCameraIdList() and getCameraIdListNoLazy(); Camera.open() hunk unchanged.

Change-Id: I9c43fd554cc0e3d477b9497028681b0e9d0afe98
---
 core/java/android/hardware/Camera.java                | 5 +++++
 core/java/android/hardware/camera2/CameraManager.java | 9 +++++++++
 2 files changed, 14 insertions(+)

diff --git a/core/java/android/hardware/Camera.java b/core/java/android/hardware/Camera.java
index 14d8a28546be..5a99bf73ddcb 100644
--- a/core/java/android/hardware/Camera.java
+++ b/core/java/android/hardware/Camera.java
@@ -518,6 +518,11 @@ public class Camera {
      * @see #open(int)
      */
     public static Camera open() {
+        int altPrimaryCamera = SystemProperties.getInt("persist.sys.alt_primary_camera", -1);
+        if (altPrimaryCamera > 0) {
+            return new Camera(altPrimaryCamera);
+        }
+
         int numberOfCameras = getNumberOfCameras();
         CameraInfo cameraInfo = new CameraInfo();
         for (int i = 0; i < numberOfCameras; i++) {
diff --git a/core/java/android/hardware/camera2/CameraManager.java b/core/java/android/hardware/camera2/CameraManager.java
index 5cd5ecd7beda..0aa66c14d20f 100644
--- a/core/java/android/hardware/camera2/CameraManager.java
+++ b/core/java/android/hardware/camera2/CameraManager.java
@@ -2568,6 +2568,15 @@ public final class CameraManager {
                             return s1.compareTo(s2);
                         }
                     }});
+
+            // HAXX: Allow overriding default primary camera (assumed to be camera 0) via property
+            // Should match with libcameraservice/common/CameraProviderManager.cpp
+            int altPrimaryCamera = SystemProperties.getInt("persist.sys.alt_primary_camera", -1);
+            if (altPrimaryCamera > 0 && altPrimaryCamera < cameraIds.length) {
+                String origPrimary = cameraIds[0];
+                cameraIds[0] = cameraIds[altPrimaryCamera];
+                cameraIds[altPrimaryCamera] = origPrimary;
+            }
         }
 
         private boolean shouldHideCamera(int currentDeviceId, int devicePolicy,
-- 
2.43.0

